compiler/rustc_metadata/src/locator.rs

```diff
@@ -723,18 +733,49 @@ impl<'a> CrateLocator<'a> {
             rejected_via_kind: self.rejected_via_kind,
             rejected_via_version: self.rejected_via_version,
             rejected_via_filename: self.rejected_via_filename,
+            rejected_via_invalid: self.rejected_via_invalid,
         })
     }
 }
 
+enum MetadataError<'a> {
+    /// The file was missing.
+    NotPresent(&'a Path),
+    /// The file was present and valid, but built by a different and incompatible version of the compiler.
+    WrongVersion(&'a Path),
```

Comment:
This one also looks legitimately useful to report.
`NotPresent` seems like the only case where we can be completely silent.

Reply:
@bjorn3 asked me not to report this, since it happens quite often (e.g. when you first build with stable and again with nightly): https://github.com/rust-lang/rust/pull/88368#discussion_r697604206. But now that the logic only reports an error if no rlibs at all were found, I agree it makes sense to report the wrong version too.

Reply:
(The commit I pushed just now has nonsensical output for this case; working now on fixing it.)